File: src/lib/types/press.ts
// Press types + input commands (mirrors src-tauri/src/config/schema/press.rs).

import type { PiePress } from "./pie";

export type InputCommand =
  | { type: "keyboard"; value: string }
  /** Mouse click; `double` fires two rapid clicks. */
  | { type: "mouseButton"; value: "left" | "right" | "middle"; double?: boolean }
  | { type: "scroll"; value: "up" | "down"; amount?: number }
  /** Reference a saved definition (input type). Fires that definition's inputs,
   * resolved live — editing the definition updates every reference. */
  | { type: "def"; def: string };

export type InputMode = "tap" | "hold" | "toggle";

export interface InputPress {
  type: "input";
  mode?: InputMode;
  label?: string;
  repeatMs?: number;
  inputs: InputCommand[];
}

/** Hold to switch to `layer`; releasing restores the previously active layer. */
export interface LayerHoldPress {
  type: "layerHold";
  label?: string;
  layer: string;
  /** Optional keys/modifiers held while the layer is active (lets a momentary
   * layer double as a combination button). */
  inputs?: InputCommand[];
}

/** A newly-added assignment with no type chosen yet. UI-only: these are never
 * written to disk (dropped on save). */
export interface NonePress {
  type: "none";
  label?: string;
}

export type PressConfig =
  | NonePress
  | InputPress
  | PiePress
  | LayerHoldPress;

export type PressType = PressConfig["type"];

File: src-tauri/src/config/schema/press.rs
//! Press types: what a button does when pressed (input / pie / layer-hold) and
//! the individual input commands a press fires.

use serde::{Deserialize, Serialize};

use crate::keys::MouseButton;

use super::*;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum PressConfig {
    Input(InputPress),
    // Boxed: `PiePress` is much larger than the other variants, so keeping it
    // inline made every stored assignment (mostly plain key inputs) that big.
    // serde treats `Box<T>` transparently, so the JSON shape is unchanged.
    Pie(Box<PiePress>),
    LayerHold(LayerHoldPress),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum InputMode {
    #[default]
    Tap,
    Hold,
    /// Press once to latch the keys down, press again to release them. The keys
    /// stay held between presses (unlike Hold, which releases on button-up).
    Toggle,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InputPress {
    #[serde(default)]
    pub mode: InputMode,
    #[serde(default)]
    pub label: Option<String>,
    #[serde(default)]
    pub repeat_ms: Option<u64>,
    #[serde(default)]
    pub inputs: Vec<InputCommand>,
}

/// Hold to switch to `layer`; releasing restores the previously active layer.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LayerHoldPress {
    #[serde(default)]
    pub label: Option<String>,
    #[serde(default)]
    pub layer: String,
    /// Optional keys/modifiers held while the layer is active (lets a momentary
    /// layer double as a combination button, e.g. also hold Ctrl).
    #[serde(default)]
    pub inputs: Vec<InputCommand>,
}

/// User-facing input descriptor (mirrors `InputCommand` in `src/lib/types.ts`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum InputCommand {
    Keyboard {
        value: String,
    },
    MouseButton {
        #[serde(default = "default_mouse_left")]
        value: MouseButtonName,
        /// Fire two rapid clicks (double-click) instead of one.
        #[serde(default)]
        double: bool,
    },
    Scroll {
        value: ScrollDir,
        #[serde(default)]
        amount: Option<i32>,
    },
    /// Reference a saved definition (input type): fires that definition's inputs.
    /// Resolved live from the definitions map so editing the definition updates
    /// every reference. The runtime expands it before firing.
    Def {
        #[serde(default)]
        def: String,
    },
}

fn default_mouse_left() -> MouseButtonName {
    MouseButtonName::Left
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MouseButtonName {
    Left,
    Right,
    Middle,
}

impl From<MouseButtonName> for MouseButton {
    fn from(v: MouseButtonName) -> Self {
        match v {
            MouseButtonName::Left => MouseButton::Left,
            MouseButtonName::Right => MouseButton::Right,
            MouseButtonName::Middle => MouseButton::Middle,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ScrollDir {
    Up,
    Down,
}
